### ISSUE TITLE
Compilation freezes

### ISSUE THREAD

**Issue by polouis** [EricssonResearch/openwebrtc#564]
Hello,

I'm trying to build openwebrtc on Mac OSX with cerbero as documented in the wiki and when entering the command :
`./cerbero-uninstalled -c config/osx-x86-64.cbc bootstrap`
the compilation hangs during this step : 
`(16/23) gobject-introspection -> compile `

What can I do to solve this? Is there some information I should provide to be more explicit about the issue?

Best regards

**Comment by superdump** [comment]
Strange, I haven't observed this. Maybe try running the build of that package on its own.

```
./cerbero-uninstalled -c config/osx-x86-64.cbc buildone gobject-introspection
```

**Comment by polouis** [comment]
Thanks for the answer. Here is the result for this command : 

```
./cerbero-uninstalled -c config/osx-x86-64.cbc buildone gobject-introspection
WARNING: Using default configuration because /Users/polouis/.cerbero/cerbero.cbc is missing
Building the following recipes: gobject-introspection
[(1/1) gobject-introspection -> fetch ]
-----> Fetching tarball http://ftp.gnome.org/pub/GNOME/sources/gobject-introspection/1.46/gobject-introspection-1.46.0.tar.xz to /Users/polouis/cerbero/sources/local/gobject-introspection-1.46.0/gobject-introspection-1.46.0.tar.xz
File /Users/polouis/cerbero/sources/local/gobject-introspection-1.46.0/gobject-introspection-1.46.0.tar.xz already downloaded.
[(1/1) gobject-introspection -> extract ]
-----> Extracting tarball to /Users/polouis/cerbero/sources/darwin_x86_64/gobject-introspection-1.46.0
Unpacking /Users/polouis/cerbero/sources/local/gobject-introspection-1.46.0/gobject-introspection-1.46.0.tar.xz in /Users/polouis/cerbero/sources/darwin_x86_64
[(1/1) gobject-introspection -> configure ]
-----> copying /Users/polouis/cerbero/data/autotools/config.guess to /Users/polouis/cerbero/sources/darwin_x86_64/gobject-introspection-1.46.0/build-aux/config.guess
-----> copying /Users/polouis/cerbero/data/autotools/config.sub to /Users/polouis/cerbero/sources/darwin_x86_64/gobject-introspection-1.46.0/build-aux/config.sub
configure: WARNING: unrecognized options: --enable-introspection
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... build-aux/install-sh -c -d
checking for gawk... no
checking for mawk... no
checking for nawk... no
checking for awk... awk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether UID '501' is supported by ustar format... yes
checking whether GID '20' is supported by ustar format... yes
checking how to create a ustar tar archive... gnutar
checking whether to enable maintainer-specific portions of Makefiles... no
checking whether make supports nested variables... (cached) yes
checking build system type... x86_64-apple-darwin12
checking host system type... x86_64-apple-darwin12
checking for gcc... clang
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether clang accepts -g... yes
checking for clang option to accept ISO C89... none needed
checking whether clang understands -c and -o together... rm: conftest.dSYM: is a directory
yes
checking for style of include used by make... GNU
checking dependency style of clang... gcc3
checking how to print strings... printf
checking for a sed that does not truncate output... /Users/polouis/cerbero/build-tools/bin/sed
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for fgrep... /usr/bin/grep -F
checking for ld used by clang... ld
checking if the linker (ld) is GNU ld... no
checking for BSD- or MS-compatible name lister (nm)... nm
checking the name lister (nm) interface... rm: conftest.dSYM: is a directory
BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 196608
checking whether the shell understands some XSI constructs... yes
checking whether the shell understands "+="... yes
checking how to convert x86_64-apple-darwin12 file names to x86_64-apple-darwin12 format... func_convert_file_noop
checking how to convert x86_64-apple-darwin12 file names to toolchain format... func_convert_file_noop
checking for ld option to reload object files... -r
checking for objdump... no
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for archiver @FILE support... rm: conftest.dSYM: is a directory
no
checking for strip... strip
checking for ranlib... ranlib
checking command to parse nm output from clang object... rm: conftest.dSYM: is a directory
ok
checking for sysroot... no
checking for mt... no
checking if : is a manifest tool... no
checking for dsymutil... dsymutil
checking for nmedit... nmedit
checking for lipo... lipo
checking for otool... otool
checking for otool64... no
checking for -single_module linker flag... yes
checking for -exported_symbols_list linker flag... yes
checking for -force_load linker flag... yes
checking how to run the C preprocessor... clang -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if clang supports -fno-rtti -fno-exceptions... yes
checking for clang option to produce PIC... -fno-common -DPIC
checking if clang PIC flag -fno-common -DPIC works... yes
checking if clang static flag -static works... no
checking if clang supports -c -o file.o... yes
checking if clang supports -c -o file.o... (cached) yes
checking whether the clang linker (ld) supports shared libraries... yes
checking dynamic linker characteristics... darwin12 dyld
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking pkg-config is at least version 0.9.0... yes
checking for flex... flex
checking lex output file root... lex.yy
checking lex library... -ll
checking whether yytext is a pointer... yes
checking for bison... bison -y
checking for dlopen in -ldl... yes
checking for the suffix of shared libraries... .dylib
checking for GLIB... no
configure: error: Package requirements (glib-2.0 >= 2.45.3) were not met:

No package 'glib-2.0' found

Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.

Alternatively, you may set the environment variables GLIB_CFLAGS
and GLIB_LIBS to avoid the need to call pkg-config.
See the pkg-config man page for more details.
Running command './configure --prefix /Library/Frameworks/OpenWebRTC.framework/Versions/0.3 --libdir /Library/Frameworks/OpenWebRTC.framework/Versions/0.3/lib  --disable-maintainer-mode  --disable-silent-rules  --enable-introspection  --build=x86_64-apple-darwin12'

Recipe 'gobject-introspection' failed at the build step 'configure'
Select an action to proceed:
[0] Enter the shell
[1] Rebuild the recipe from scratch
[2] Rebuild starting from the failed step
[3] Skip recipe
[4] Abort

```
That's weird since glib has been build during the whole process. The PKG_CONFIG_PATH environment variable is not set and I don't know which path should be set.

**Comment by stefanalund** [comment]
I'm getting a similar stall when compiling `gst-plugins-bad-1.0`. Following @superdump's suggestion above (`./cerbero-uninstalled -c config/osx-x86-64.cbc buildone gst-plugins-bad-1.0`) does not change things.

**Comment by stefhak** [comment]
I have exactly the same issue as @stefanalund

**Comment by superdump** [comment]
I'll try a clean build to see if I can reproduce. The steps I will take are:

```
cd ~/cerbero

./cerbero-uninstall -c config/osx-x86-64.cbc wipe --build-tools
./cerbero-uninstall -c config/cross-ios-universal.cbc wipe --build-tools

sudo mkdir -p /Library/Frameworks/OpenWebRTC.framework
sudo chown -R $UID /Library/Frameworks/OpenWebRTC.framework/

mkdir -p dist
mkdir -p /Library/Frameworks/OpenWebRTC.framework/Versions/0.3
ln -sf /Library/Frameworks/OpenWebRTC.framework/Versions/0.3 dist/darwin_x86_64

./cerbero-uninstall -c config/osx-x86-64.cbc fetch-package --full-reset --reset-rdeps openwebrtc && \
./cerbero-uninstall -c config/cross-ios-universal.cbc fetch-package --full-reset --reset-rdeps openwebrtc && \
./cerbero-uninstall -c config/osx-x86-64.cbc bootstrap && \
./cerbero-uninstall -c config/cross-ios-universal.cbc bootstrap && \
./cerbero-uninstall -c config/osx-x86-64.cbc package openwebrtc && \
./cerbero-uninstall -c config/cross-ios-universal.cbc package openwebrtc
```

...for the targets you're interested in. Say 'yes' to removing directories in the first two wipe steps.

**Comment by stefhak** [comment]
s/cerbero-uninstall/cerbero-uninstalled/

**Comment by superdump** [comment]
The OS X package step has completed flawlessly so now it's just whether the universal iOS package step completes or not... It seems I cannot reproduce the issue.

**Comment by superdump** [comment]
And now the iOS build completed in full and without error.

**Comment by SrujithPoondla** [comment]
The same issue I am getting in Linux @superdump 
The compilation freezes at the gobject-introspection install step.

**Comment by stefanalund** [comment]
With the instructions provided by @superdump above I was now able to build everyting again :+1: 

Maybe we should add those instructions to the Wiki?

**Comment by stefhak** [comment]
Building worked fine for me too now, thanks @superdump !
But running the daemon gives:
```
** (gst-plugin-scanner:94266): CRITICAL **: can't find gi.repository.Gst

** (gst-plugin-scanner:94266): CRITICAL **: can't find gi.repository.Gst
{"line":24,"column":32,"sourceURL":"/Library/Frameworks/OpenWebRTC.framework/Versions/0.3/share/seed-gtk3/extensions/Seed.js"}
Origin of extension incorrect, or a socket to the extension did already exist; closing
```

**Comment by superdump** [comment]
@stefhak - did you run openwebrtc-daemon in the cerbero shell?

**Comment by stefhak** [comment]
@superdump yes, it was run in the shell:
```
$ ./cerbero-uninstalled shell
WARNING: Using default configuration because /Users/eplhak/.cerbero/cerbero.cbc is missing
[cerbero-darwin-x86_64]:~/cerbero$ openwebrtc-daemon 

** (gst-plugin-scanner:94266): CRITICAL **: can't find gi.repository.Gst

** (gst-plugin-scanner:94266): CRITICAL **: can't find gi.repository.Gst
{"line":24,"column":32,"sourceURL":"/Library/Frameworks/OpenWebRTC.framework/Versions/0.3/share/seed-gtk3/extensions/Seed.js"}
Origin of extension incorrect, or a socket to the extension did already exist; closing
```

**Comment by superdump** [comment]
And you are running this on an OS X machine, right?

**Comment by stefhak** [comment]
Yes, BTW
```
[cerbero-darwin-x86_64]:~/cerbero$ which openwebrtc-daemon
/Users/myusername/cerbero/dist/darwin_x86_64/bin/openwebrtc-daemon
[cerbero-darwin-x86_64]:~/cerbero$ 
```

**Comment by superdump** [comment]
```
ls -l ~/cerbero/dist
G_DEBUG=fatal-criticals lldb openwebrtc-daemon
```

Then type `r` and hit enter at the prompt. It should drop to the prompt at the first critical. Then run `bt` and paste the output to a gist or pastebin and link it here.

**Comment by stefhak** [comment]
Here it is:
```
[cerbero-darwin-x86_64]:~/cerbero$ G_DEBUG=fatal-criticals lldb openwebrtc-daemon
(lldb) target create "openwebrtc-daemon"
Current executable set to 'openwebrtc-daemon' (x86_64).
(lldb) r
Process 94819 launched: '/Users/eplhak/cerbero/dist/darwin_x86_64/bin/openwebrtc-daemon' (x86_64)
{"line":24,"column":32,"sourceURL":"/Library/Frameworks/OpenWebRTC.framework/Versions/0.3/share/seed-gtk3/extensions/Seed.js"}
Origin of extension incorrect, or a socket to the extension did already exist; closing
**
GLib:ERROR:gmain.c:6319:cocoa_select_thread_collect_poll: assertion failed: (ufds[i].fd == cocoa_current_pollfds[i].fd)
Process 94819 stopped
* thread #1: tid = 0x7ee0fb, 0x00007fff9c2be002 libsystem_kernel.dylib`__pthread_kill + 10, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x00007fff9c2be002 libsystem_kernel.dylib`__pthread_kill + 10
libsystem_kernel.dylib`__pthread_kill:
->  0x7fff9c2be002 <+10>: jae    0x7fff9c2be00c            ; <+20>
    0x7fff9c2be004 <+12>: movq   %rax, %rdi
    0x7fff9c2be007 <+15>: jmp    0x7fff9c2b8bdd            ; cerror_nocancel
    0x7fff9c2be00c <+20>: retq   
(lldb) bt
* thread #1: tid = 0x7ee0fb, 0x00007fff9c2be002 libsystem_kernel.dylib`__pthread_kill + 10, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x00007fff9c2be002 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff8ee3a5c5 libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fff93f0a6e7 libsystem_c.dylib`abort + 129
    frame #3: 0x0000000100b37fc3 libglib-2.0.0.dylib`g_assertion_message(domain=<unavailable>, file=<unavailable>, line=<unavailable>, func=<unavailable>, message=<unavailable>) + 387 at gtestutils.c:2429
    frame #4: 0x0000000100b3802e libglib-2.0.0.dylib`g_assertion_message_expr(domain="GLib", file="gmain.c", line=6319, func="cocoa_select_thread_collect_poll", expr=<unavailable>) + 94 at gtestutils.c:2444
    frame #5: 0x0000000100b14cf5 libglib-2.0.0.dylib`cocoa_select_thread_collect_poll(ufds=<unavailable>, nfds=<unavailable>) + 181 at gmain.c:6319
    frame #6: 0x0000000100b11c25 libglib-2.0.0.dylib`cocoa_poll_func(ufds=0x00000001016303f0, nfds=5, timeout_=0) + 213 at gmain.c:6387
    frame #7: 0x0000000100b12c6d libglib-2.0.0.dylib`g_main_context_iterate [inlined] g_main_context_poll(priority=0) + 41 at gmain.c:4264
    frame #8: 0x0000000100b12c44 libglib-2.0.0.dylib`g_main_context_iterate(context=<unavailable>, block=<unavailable>, dispatch=<unavailable>, self=<unavailable>) + 292 at gmain.c:3964
    frame #9: 0x0000000100b12fef libglib-2.0.0.dylib`g_main_loop_run(loop=0x000000010225a5e0) + 223 at gmain.c:4163
    frame #10: 0x0000000100048285 libopenwebrtc.4201.dylib`owr_run + 53 at owr.c:359
    frame #11: 0x0000000100013a8b libopenwebrtc_bridge.0.dylib`run(msg_queue=<unavailable>) + 1179 at owr_bridge.c:422
    frame #12: 0x0000000100008f49 openwebrtc-daemon`main(argc=<unavailable>, argv=<unavailable>) + 9 at daemon.c:37
    frame #13: 0x00007fff8c26f5ad libdyld.dylib`start + 1
(lldb) 
```

**Comment by stefhak** [comment]
BTW, tried @pererikb patch (added ```ensureWebSocket();``` on empty line 87 of webrtc.js) and things changed to:
```
$ openwebrtc-daemon 
{"line":24,"column":32,"sourceURL":"/Library/Frameworks/OpenWebRTC.framework/Versions/0.3/share/seed-gtk3/extensions/Seed.js"}
Origin of extension incorrect, or a socket to the extension did already exist; closing
**
GLib:ERROR:gmain.c:6319:cocoa_select_thread_collect_poll: assertion failed: (ufds[i].fd == cocoa_current_pollfds[i].fd)
Abort trap: 6
```
@superdump @pererikb any theories?

**Comment by superdump** [comment]
Are you sure it is the only openwebrtc-daemon running? Is anything else binding the socket that it wants to use?

**Comment by stefhak** [comment]
Restarting Safari fixed this problem. Perhaps we can now close this Issue? Building is no problem (at least not for me) for the moment.

**Comment by superdump** [comment]
Closing. Follow the steps above to do a clean build if you encounter these kinds of issues.

**Comment by polouis** [comment]
I'm sorry to answer this late but the clean build commands @superdump provided did not solve the issue for me. I noticed that if I build any package on its own, it fails with the error about PKG_CONFIG_PATH, even for packages that are built with success when running the whole build process. That means that the issue I have when building a package alone may not be the same as the one I have when running the whole build process.

**Comment by superdump** [comment]
@polouis - which platform?

**Comment by polouis** [comment]
Mac OS X

**Comment by superdump** [comment]
@polouis - which version and which version of Xcode?

**Comment by polouis** [comment]
Hello,
I have Version 7.2.1 (7C1002) installed on OS X Yosemite 10.10.5 (14F1605).
I retried the command ./cerbero-uninstalled -c config/osx-x86-64.cbc buildone gobject-introspection
because I did not understand first that when prompted with the error about PKG_CONFIG_PATH I had to choose option "[0] Enter the shell" to examine environment variables. So, I have that : 
```
declare -x PKG_CONFIG="/Users/polouis/cerbero/build-tools/bin/pkg-config"
declare -x PKG_CONFIG_LIBDIR="/Library/Frameworks/OpenWebRTC.framework/Versions/0.3/lib/pkgconfig"
declare -x PKG_CONFIG_PATH="/Library/Frameworks/OpenWebRTC.framework/Versions/0.3/share/pkgconfig"
```
which seems OK. But when I do a ls -R into the framework directory I get : 
```
Versions
./Versions:
0.3
./Versions/0.3:
0.3	include	share
./Versions/0.3/include:
GL
./Versions/0.3/share:
aclocal
./Versions/0.3/share/aclocal:
```
So, what I understand is that I get the error "No package 'glib-2.0' found" because even if glib was compiled successfully during a previous step, some files were not copied to framework directory. Am I wrong?
